File: 2017/day05-1.hs
solution :: String -> Int
solution = play 0 0 . map read . lines

play :: Int -> Int -> [Int] -> Int
play step index jumps 
    | index < 0 || index >= length jumps = step
    | otherwise = play (step +1) (index + (!!index) jumps) (take index jumps ++ [(!!index) jumps + 1] ++ drop (index+1) jumps) 

main :: IO ()
main = do readFile "day05.dat" >>= print . solution

File: 2017/day05_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    return play([int(j) for j in content])


def play(jumps: list[int]):
    step = 0
    index = 0
    copy = jumps[::]
    while 0 <= index < len(copy):
        offset = copy[index]
        copy[index] += 1
        index += offset
        step += 1
    return step


def main() -> None:
    with open("day05.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
